src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -229,6 +230,10 @@ internal Binder GetSpeculativeBinder(int position, ExpressionSyntax expression,
                 //Debug.Assert(binder is TypeofBinder); // Expectation, not requirement.
                 binder = new TypeofBinder(expression, binder);
             }
+            else if (bindingOption == SpeculativeBindingOption.BindAsExpression && IsInNameofExpression(position, binder, out var argumentExpression))
+            {
+                binder = new NameofBinder(argumentExpression, binder);
```

Comment:
I haven't reviewed other paths where we may need `NameofBinder`. The PR is just a quick prototype to get initial feedback

Reply:
I added more tests hoping to catch other missing `NameofBinder`s, but none of the tests caught anything. Let me know if you think it's missing somewhere.